README.md
three-software-renderer
=======================
##Universal Three.js in-memory renderer for node.js and the browser

# What's this for?
The `SoftwareRenderer` is a drop-in renderer for [Three.js](http://www.threejs.org) that can be used on the server and the client,
if you need to be GPU independent.

# What's the drawback?
Primarily performance. As this renderer does not have access to GPU acceleration and GPU shader units, it's much much slower than the CanvasRenderer or WebGLRenderer.
In addition, you'll have to use `THREE.DataTexture` instead of a normal `THREE.Texture` for textures when you're in a headless environment (i.e. node.js).

# What are possible use cases?
* Rendering in a web worker, e.g. for static renderings that should happen in the background
* Rendering on the server, e.g. for preview images of 3D models
* Tell me, if you have other ideas

# Which Three.js version is this compatible with?
So far I have tested 0.69 and 0.71. Most recently also tested with three.js v0.112.1. Please tell me if you tried it with other versions as well :-)

# What is this work based on?
First and foremost this is just a modified copy of the SoftwareRenderer found in [the Three.js example sources](https://github.com/mrdoob/three.js/blob/0b07813dc45481f1d16d3b6d2334178664861465/examples/js/renderers/SoftwareRenderer.js).
It also uses the modified [THREE.Projector](https://github.com/mrdoob/three.js/blob/20b77b2785afaa4d00a1ecd222e6de1a3ec76006/examples/js/renderers/Projector.js) from the examples.

# How to use
```javascript
const THREE = require("three");
const SoftwareRenderer = require("three-software-renderer");
const PNG = require("pngjs").PNG;
const fs = require("fs");

// Build scene with cube
const width = 1024;
const height = 768;

const camera = new THREE.PerspectiveCamera(75, width / height, 1, 10000);
camera.position.z = 500;

// Create a simple red cube
const geometry = new THREE.BoxGeometry(200, 200, 200);
const material = new THREE.MeshBasicMaterial({color: 0xff0000});
const mesh = new THREE.Mesh(geometry, material);

// Rotate the cube a bit
mesh.rotation.x += 0.5;
mesh.rotation.y += 0.6;

const scene = new THREE.Scene();
scene.add(mesh);

// Render into an Uint8ClampedArray (RGBA).
const renderer = new SoftwareRenderer({
  alpha: false
});
renderer.setSize(width, height);
var imagedata = renderer.render(scene, camera);

// Create a PNG from the array
const png = new PNG({
  width: width,
  height: height,
  filterType: -1
});

// Copy byte array to png object  
for(var i=0;i<imagedata.data.length;i++) {
  png.data[i] = imagedata.data[i];
}

// Write PNG to disk
if (!fs.existsSync("temp")) {
  fs.mkdirSync("temp");
}
png.pack().pipe(fs.createWriteStream("temp/example.png"));
```

# I found a bug, how can I help?
Please help me improve this module by opening an issue on Github. If you have an idea what might cause the problem, I'd be very grateful if you'd try to investigate the bug and make a pull request to this repository.
If you have any questions or need help doing so, let me know via the issues in this project!


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added
- Added `CHANGELOG.md`
- Added `package-lock.json` as a repo tracked file
- Added `three` NPM to **peerDependencies**
- Added **engines** property to `package.json`
- Added `three-ply-loader` as **devDependencies** for an additional example user case (NodeJS wrapper for Three.js' PLYLoader function)
- Added `examples/ply-model-to-png.js` source code example
- Added `examples/assets/cube.ply` simple PLY 3D model
- Added `.nvmrc` in order to easily support [node version manager][nvm]


### Changed
- Upgraded `three` NPM to v0.112.1
- Upgraded `mocha` NPM to v6.2.2
- Upgraded `pngjs` NPM to v3.4.0
- Moved `three` NPM to **devDependencies** in `package.json`
- Dropped `.applyProjection()` in favour of `.applyMatrix4(m)` in `src/projector.js` source file
- Upgraded `src/projector.js` to the latest three.js [source file][three-js-projector-source-link] available in order to encompass when **material.vertexColors** equals:
  - `THREE.VertexColors`
  - `THREE.FaceColors`
- Minor edit in `src/software-renderer.js` in order to encompass when **material.vertexColors** is set to `THREE.VertexColors`
- Renamed `example` folder to `examples`
- Moved `checkerboard.png` to `examples/assets` folder
- Minor path edit in `examples/texture.js`

[three-js-projector-source-link]: https://github.com/mrdoob/three.js/blob/dev/examples/js/renderers/Projector.js
[nvm]: https://github.com/nvm-sh/nvm

example/main.js
/**
 * @author jonaskello / https://github.com/jonaskello
 * Simple example that renders a scene with a cube to a PNG image file.
 */
const THREE = require("three");

// Use either named export or default export
//const SoftwareRenderer = require("../");
const SoftwareRenderer = require("../").SoftwareRenderer;

const PNG = require("pngjs").PNG;
const fs = require("fs");

// Build scene with cube
const width = 1024;
const height = 768;
const camera = new THREE.PerspectiveCamera(75, width / height, 1, 10000);
camera.position.z = 500;
const scene = new THREE.Scene();
const geometry = new THREE.BoxGeometry(200, 200, 200);
const material = new THREE.MeshBasicMaterial({color: 0xff0000});
const mesh = new THREE.Mesh(geometry, material);
scene.add(mesh);

// Rotate the cube a bit
mesh.rotation.x += 0.5;
mesh.rotation.y += 0.6;

// Render into pixels-array (RGBA)
const renderer = new SoftwareRenderer();
renderer.setSize(width, height);
var imagedata = renderer.render(scene, camera);

// Create a PNG from the pixels array (RGBA)
const png = new PNG({
  width: width,
  height: height,
  filterType: -1
});

for(var i=0;i<imagedata.data.length;i++) {
  png.data[i] = imagedata.data[i];
}
console.log(png.data);
if (!fs.existsSync("temp")) {
  fs.mkdirSync("temp");
}
png.pack().pipe(fs.createWriteStream("temp/example.png"));
